luigi/contrib/opener.py

```diff
@@ -0,0 +1,270 @@
+        index = self.registry[name]
+        return self.openers[index]
+
+    def add(self, opener):
+        """Adds an opener to the registry
+
+        :param opener: Opener object
+        :type opener: Opener inherited object
+
+        """
+
+        index = len(self.openers)
+        self.openers[index] = opener
+        for name in opener.names:
+            self.registry[name] = index
+
+    def open(self, target_uri, **kwargs):
+        """Open target uri.
+
+        :param target_uri: Uri to open
+        :type target_uri: string
+
+        :returns: Target object
+
+        """
+        target = urlsplit(target_uri, scheme=self.default_opener)
+
+        opener = self.get_opener(target.scheme)
+        query = opener.conform_query(target.query)
+
+        target = opener.get_target(
+            target.scheme,
+            target.path,
+            target.fragment,
+            target.username,
+            target.password,
+            target.hostname,
+            target.port,
+            query,
+            **kwargs
+        )
+        target.opener_path = target_uri
+
+        return target
+
+
+class Opener(object):
+
+    """Base class for Opener objects.
+
+    """
+
+    # Dictionary of expected kwargs and flag for json loading values (bool/int)
+    allowed_kwargs = {}
+    # Flag to filter out unexpected kwargs
+    filter_kwargs = True
+
+    @classmethod
+    def conform_query(cls, query):
+        """Converts the query string from a target uri, uses
+        cls.allowed_kwargs, and cls.filter_kwargs to drive logic.
+
+        :param query: Unparsed query string
+        :type query: urllib.parse.unsplit(uri).query
+        :returns: Dictionary of parsed values, everything in cls.allowed_kwargs
+            with values set to True will be parsed as json strings.
+
+        """
+        query = parse_qs(query, keep_blank_values=True)
+
+        # Remove any unexpected keywords from the query string.
+        if cls.filter_kwargs:
+            query = {x: y for x, y in query.items() if x in cls.allowed_kwargs}
+
+        for key, vals in query.items():
+            # Multiple values of the same name could be passed use first
+            # Also params without strings will be treated as true values
+            if cls.allowed_kwargs.get(key, False):
+                val = json.loads(vals[0] or 'true')
+            else:
+                val = vals[0] or 'true'
+
+            query[key] = val
+
+        return query
+
+    @classmethod
+    def get_target(cls, scheme, path, fragment, username,
+                   password, hostname, port, query, **kwargs):
+        """Override this method to use values from the parsed uri to initialize
+        the expected target.
+
+        """
+        raise NotImplemented("get_target must be overridden")
+
+
+class MockOpener(Opener):
+
+    """Mock target opener, works like LocalTarget but files are all in
+    memory.
+
+    example:
+    * mock://foo/bar.txt
```

Comment:
minor thing but shouldn't it be just "mock:foo/bar.txt"? double slash usually indicates a hostname

not blocking